flink/src/main/java/org/apache/iceberg/flink/source/IcebergSource.java

```diff
@@ -0,0 +1,207 @@
+
+  private final TableLoader tableLoader;
+  private final ScanContext scanContext;
+  private final BulkFormat<T, IcebergSourceSplit> bulkFormat;
+  private final SplitAssignerFactory assignerFactory;
+  private final IcebergEnumeratorConfig enumeratorConfig;
+
+  IcebergSource(
+      TableLoader tableLoader,
+      ScanContext scanContext,
+      BulkFormat<T, IcebergSourceSplit> bulkFormat,
+      SplitAssignerFactory assignerFactory,
+      IcebergEnumeratorConfig enumeratorConfig) {
+
+    this.tableLoader = tableLoader;
+    this.enumeratorConfig = enumeratorConfig;
+    this.scanContext = scanContext;
+    this.bulkFormat = bulkFormat;
+    this.assignerFactory = assignerFactory;
+  }
+
+  private static Table loadTable(TableLoader tableLoader) {
+    tableLoader.open();
+    try (TableLoader loader = tableLoader) {
+      return loader.loadTable();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to close table loader", e);
+    }
+  }
+
+  @Override
+  public Boundedness getBoundedness() {
+    return enumeratorConfig.splitDiscoveryInterval() == null ?
+        Boundedness.BOUNDED : Boundedness.CONTINUOUS_UNBOUNDED;
+  }
+
+  @Override
+  public SourceReader<T, IcebergSourceSplit> createReader(SourceReaderContext readerContext) {
+    return new IcebergSourceReader<>(
+        readerContext,
+        bulkFormat);
+  }
+
+  @Override
+  public SplitEnumerator<IcebergSourceSplit, IcebergEnumeratorState> createEnumerator(
+      SplitEnumeratorContext<IcebergSourceSplit> enumContext) {
+    return createEnumerator(enumContext, null);
+  }
+
+  @Override
+  public SplitEnumerator<IcebergSourceSplit, IcebergEnumeratorState> restoreEnumerator(
+      SplitEnumeratorContext<IcebergSourceSplit> enumContext, IcebergEnumeratorState enumState) {
+    return createEnumerator(enumContext, enumState);
+  }
+
+  @Override
+  public SimpleVersionedSerializer<IcebergSourceSplit> getSplitSerializer() {
+    return IcebergSourceSplitSerializer.INSTANCE;
+  }
+
+  @Override
+  public SimpleVersionedSerializer<IcebergEnumeratorState> getEnumeratorCheckpointSerializer() {
+    return IcebergEnumeratorStateSerializer.INSTANCE;
+  }
+
+  private SplitEnumerator<IcebergSourceSplit, IcebergEnumeratorState> createEnumerator(
+      SplitEnumeratorContext<IcebergSourceSplit> enumContext,
+      @Nullable IcebergEnumeratorState enumState) {
+
+    final Table table = loadTable(tableLoader);
+
+    final SplitAssigner assigner;
+    if (enumState == null) {
+      assigner = assignerFactory.createAssigner();
+      // for batch jobs, discover splits eagerly during job initialization.
+      // As FLINK-16866 supports non-blocking job submission since 1.12,
+      // heavy job initialization won't lead to request timeout for job submission.
+      if (enumeratorConfig.splitDiscoveryInterval() == null) {
+        assigner.onDiscoveredSplits(FlinkSplitGenerator.planIcebergSourceSplits(table, scanContext));
```

Comment:
I haven't run into problems with unit test or when running it in our infrastructure at Apple. But I can definitely see your point. Instead of passing `Table`, I will pass the `TableLoader` to avoid the potential problem you mentioned